File: coinChange/coinChange.go
// https://leetcode.com/problems/coin-change/
package coinchange

func coinChange(coins []int, amount int) int {
	dp := make([]int, amount+1)
	for i := range dp {
		dp[i] = amount + 1 // not found
	}
	dp[0] = 0 // base case

	for _, c := range coins {
		for a := c; a <= amount; a++ {
			if (a - c) >= 0 {
				dp[a] = min(dp[a], 1+dp[a-c])
			}
		}
	}

	// combination not found
	if dp[amount] > amount {
		return -1
	}

	return dp[amount]
}

File: neet/coin_change.py
# https://neetcode.io/problems/coin-change

class Solution:
    def coinChange(self,coins:list[int], amount:int)->int:
        dp = [amount+1]*(amount+1)
        dp[0] = 0
        for a in range(1,amount+1):
            for c in coins:
                if a-c>=0:
                    dp[a] = min(dp[a],dp[a-c]+1)
        return dp[amount] if dp[amount] != amount+1 else -1


s = Solution()

coins = [1,5,10]
amount = 12
assert s.coinChange(coins, amount) == 3

coins = [2]
amount = 3
assert s.coinChange(coins, amount) == -1

coins = [1]
amount = 0
assert s.coinChange(coins, amount) == 0
